### PR TITLE
th/ip6_temp_addr_sync_rh1542609

### PR DESCRIPTION
**Pull request by thom311**
Fix syncing IPv6 addresses

(the last commit is just some extra verbose logging for debugging. Ignore that one).

See also: https://bugzilla.redhat.com/show_bug.cgi?id=1542609

### CODE REVIEW

**Comment by lkundrak** [comment]
> platform: add nm_platform_refresh_all() API
...
> It's only use is for testing
> really, because if there is a known condition in which the cache gets
> out of sync, that should be fixed instead.

Why include the patch then? Please elaborate on what is the "known condition"
because it is not known to me.


> platform: maintain proper order of addresses in platform cache during RTM_NEWADDR
...
>+       /* new objects are usually appended to the list. Except for
>+        * addresses, which are prepended.

Why?

> TEST

I guess you're not going to merge this one?

**Comment by thom311** [comment]
> >   It's only use is for testing
> >   really, because if there is a known condition in which the cache gets
> >   out of sync, that should be fixed instead.
>
> Why include the patch then? Please elaborate on what is the "known condition"
> because it is not known to me.

Reworded commit message. It is a useful function like nm_ip4_config_dump(), which is never used in production code, but it's useful to have around during a printf debugging session. Same here, that is why I would merge the patch.



> >          * addresses, which are prepended.
>
> Why?

Improved(?) comment




> >    TEST
>
> I guess you're not going to merge this one?

Correct. I drop this one.


Repushed.

**Comment by lkundrak** [comment]
LGTM

File: src/platform/nm-platform.c [hunk 1]
```diff
@@ -3545,17 +3512,16 @@ gboolean
...
+	gs_unref_hashtable GHashTable *known_addresses_idx = NULL;
 	NMPLookup lookup;
 	guint32 ifa_flags;
-	gboolean remove = FALSE;
 
-	if (!_addr_array_clean_expired (AF_INET6, ifindex, known_addresses, now, NULL))
+	if (!_addr_array_clean_expired (AF_INET6, ifindex, known_addresses, now, &known_addresses_idx))
 		known_addresses = NULL;
 
 	/* @plat_addresses and @known_addresses are in increasing priority order */
```
Comment (bengal) [inline src/platform/nm-platform.c]: The comment seems now obsolete, as the commit messages says:

> Note that @known_addresses contains the desired addresses in order of
    least-important first, while @plat_addresses contains them in opposite
    order.

**Comment by bengal** [comment]
Otherwise LGTM

**Comment by bengal** [comment]
The following crash was caught by CI running the th/ip6-temp-addr-sync-rh1542609 branch:

https://bugzilla.redhat.com/show_bug.cgi?id=1543640

**Comment by bengal** [comment]
New commits look good.

**Comment by thom311** [comment]
I merged this to master early, because the branch got some review.

I hurry on this one, because in my opinion the issue is quite severe (it makes IPv6 unusable for me at home).

I think this should make it to master/nm-1-10 as soon as possible, so that it gets early exposure to more testing and potential remaining issues can be found sooner. This issue should be fixed ASAP and there is almost no time left for including it in rhel-7.5. If we are going to include it, it must get early testing before it's 100% perfect.

**Pull request closed without merging**